theano/tensor/nnet/nnet.py

```diff
@@ -465,119 +462,114 @@ def c_code_template(dtype):
-                     "failed to alloc sm output");
-                %(fail)s
+            // Allocate memory for the softmax output
+            if (NULL == %(sm)s || (PyArray_DIMS(%(x)s) != PyArray_DIMS(%(sm)s)))
+            {
+                Py_XDECREF(%(sm)s);
+                %(sm)s = (PyArrayObject*)PyArray_SimpleNew(ndim, PyArray_DIMS(%(x)s),
+                                                        PyArray_TYPE(%(x)s));
+                if(!%(sm)s) {
+                    PyErr_SetString(PyExc_MemoryError,
+                        "Failed to allocate memory for sm output");
+                    %(fail)s
+                }
             }
-        }
-        Sx1 = PyArray_STRIDES(%(x)s)[1]/sizeof(dtype_%(x)s);
-        Ssm1 = PyArray_STRIDES(%(sm)s)[1]/sizeof(dtype_%(sm)s);
+
+            // Compute array size except on the last dimension
+            npy_int input_size;
+            input_size = PyArray_SIZE(%(x)s) / input_shape[ndim - 1];
         """
 
         begin_row_loop = """
-        for (size_t i = 0; i < Nx[0]; ++i)
-        {
-            size_t j;
-            double sum = 0.0;
-
-            const dtype_%(x)s* __restrict__ x_i = (dtype_%(x)s*)(PyArray_BYTES(%(x)s) + PyArray_STRIDES(%(x)s)[0] * i);
-            dtype_%(sm) s* __restrict__ sm_i = (dtype_%(sm)s*)(PyArray_BYTES(%(sm)s) + PyArray_STRIDES(%(sm)s)[0] * i);
+            // Compute stride
+            npy_int stride;
+            stride = PyArray_DIM(%(x)s, ndim - 1) * PyArray_ITEMSIZE(%(x)s);
 
-            dtype_%(sm)s row_max = x_i[0];
-            //std::cout << "0 " << row_max << "\\n";
-            // Get the maximum value of the row
-            for (j = 1; j < Nx[1]; ++j)
+            for (size_t i = 0; i < input_size ; i++)
             {
-                dtype_%(sm)s row_ij = x_i[j * Sx1] ;
-                //std::cout << "1 " << row_ij << "\\n";
-                row_max   = (row_ij > row_max) ? row_ij : row_max;
-            }
-
+                size_t j;
+                double sum = 0.0;
+
+                const dtype_%(x)s* __restrict__ x_i = (dtype_%(x)s*)(PyArray_BYTES(%(x)s) + stride * i);
+                dtype_%(sm) s* __restrict__ sm_i = (dtype_%(sm)s*)(PyArray_BYTES(%(sm)s) + stride * i);
+                dtype_%(sm)s row_max = x_i[0];
+
+                // Get the maximum value on this dimenson
+                for (j = 1; j < input_shape[ndim - 1]; ++j)
+                {
+                    dtype_%(sm)s row_ij = x_i[j];
+                    row_max = (row_ij > row_max) ? row_ij : row_max;
+                }
         """
 
         inside_row_loop = """
-            for (j = 0; j < Nx[1]; ++j)
+            // Substract the max and take the exponential
+            for (j = 0; j < input_shape[ndim - 1]; ++j)
             {
-                dtype_%(sm)s row_ij = x_i[j * Sx1] ;
-                //std::cout << "2 " << j << " " << row_ij << " " << row_max << "\\n";
+                dtype_%(sm)s row_ij = x_i[j] ;
```

Comment:
In this case, we are on the last dimension. We don't really need to have a stride here since it's value will be always 1. But I agree in other cases, we should compute it.